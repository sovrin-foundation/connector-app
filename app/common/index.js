// @flow
export * from './button-constants'
export * from './message-constants'
export * from './route-constants'
export * from './secure-storage-constants'
export function noop() {}
