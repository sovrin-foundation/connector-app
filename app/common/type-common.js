// @flow

import type { Element } from 'react'
import type { NavigationScreenProp } from 'react-navigation'

export type CustomError = {
  code: string,
  message: string,
}

export const INITIAL_TEST_ACTION: 'INITIAL_TEST_ACTION' = 'INITIAL_TEST_ACTION'

export type InitialTestAction = {
  type: typeof INITIAL_TEST_ACTION,
}

export const initialTestAction = () => ({
  type: INITIAL_TEST_ACTION,
})

export type NavigationParams = {
  [string]: any,
}
export type NavigationRoute = NavigationLeafRoute
export type NavigationState = {
  index: number,
  routes: Array<NavigationRoute>,
}

export type NavigationLeafRoute = {|
  key: string,
  routeName: string,
  path?: string,
  params?: NavigationParams,
|}

export type ReactNavigation = NavigationScreenProp

export type GenericObject = {
  [string]: any,
}

export type GenericStringObject = {
  [string]: string,
}

export type MessageAnnotation = {
  name: string,
  version: string,
}

export type TopicAnnotation = {
  tid: number,
  mid: number,
}

export type ReactChildren = Element<*>

export const RESET: 'RESET' = 'RESET'
export const REMOVE_SERIALIZED_CLAIM_OFFERS_SUCCESS: 'REMOVE_SERIALIZED_CLAIM_OFFERS_SUCCESS' =
  'REMOVE_SERIALIZED_CLAIM_OFFERS_SUCCESS'

export type ResetAction = {
  type: typeof RESET,
}

export type ImageSource = {
  uri: string,
}

export type NotificationPayload = {
  forDID: string,
  uid: string,
  type: string,
  remotePairwiseDID: string,
  senderLogoUrl?: ?string,
}

export type ImmutableGenericStringObject = {
  +[string]: string,
}

export const STORE_STATUS = {
  IDLE: 'IDLE',
  IN_PROGRESS: 'IN_PROGRESS',
  ERROR: 'ERROR',
  SUCCESS: 'SUCCESS',
}

export type StoreError = { +error: ?CustomError }
export type StoreStatus = { +status: $Keys<typeof STORE_STATUS> }

export type StatusBarStyle = 'default' | 'light-content' | 'dark-content'
