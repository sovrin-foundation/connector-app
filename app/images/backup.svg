<svg width="24" height="24" xmlns="http://www.w3.org/2000/svg"><path d="M12 6v6.429l4.272 4.299-1.044 1.044-4.728-4.7V6H12zm6 1.5h3.482a10.598 10.598 0 0 0-3.877-4.373 10.569 10.569 0 0 0-2.665-1.205A10.31 10.31 0 0 0 12 1.5c-.964 0-1.893.125-2.786.375s-1.727.603-2.504 1.058A10.537 10.537 0 0 0 4.58 4.58a10.536 10.536 0 0 0-1.647 2.13 10.558 10.558 0 0 0-1.058 2.504c-.25.893-.375 1.822-.375 2.786s.125 1.893.375 2.786.603 1.727 1.058 2.504a10.537 10.537 0 0 0 1.647 2.13 10.536 10.536 0 0 0 2.13 1.647c.777.455 1.611.808 2.504 1.058s1.822.375 2.786.375c1.17 0 2.3-.19 3.388-.57a10.59 10.59 0 0 0 2.967-1.586 10.686 10.686 0 0 0 2.31-2.431 10.185 10.185 0 0 0 1.447-3.114l1.446.402a11.703 11.703 0 0 1-1.654 3.562 12.025 12.025 0 0 1-6.027 4.594A11.756 11.756 0 0 1 12 24c-1.098 0-2.16-.143-3.188-.429a12.139 12.139 0 0 1-2.872-1.205 11.9 11.9 0 0 1-2.431-1.875 11.9 11.9 0 0 1-1.875-2.43 12.168 12.168 0 0 1-1.205-2.867A11.787 11.787 0 0 1 0 12.007c0-1.103.143-2.168.429-3.195A12.139 12.139 0 0 1 1.634 5.94a11.9 11.9 0 0 1 1.875-2.431 11.9 11.9 0 0 1 2.43-1.875c.89-.518 1.844-.92 2.867-1.205A11.787 11.787 0 0 1 11.993 0c1.103 0 2.17.145 3.201.435a12.11 12.11 0 0 1 5.39 3.188c.75.763 1.39 1.627 1.916 2.591V3H24v6h-6V7.5z" fill="#777" fill-rule="nonzero"/></svg>