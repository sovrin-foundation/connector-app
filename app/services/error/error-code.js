// @flow

// We need to decide on format for error codes
// there is a convention in HTTP error codes is that all errors
// starting from 4xx are client side errors and 5xx are server side errors
// we also need to have codes in some logical sequence
// Here is the proposal for that sequence of Error codes
// 1xxxx = data validation error
// where first two `xx` denotes the area of code in which error occurred
// and third x denotes which functionality of that particular area
// and last x denotes which parameter failed
// 2xxxx = libindy exceptions
// 3xxxx = something else
// ----------
// There can another proposal where we just write a detailed error message
// and assign any type of sequential number to error codes
// error code are still needed because we need to take different actions
// for different error codes and we need to match codes, we can do string
// comparisons as well, but tracking code will be bit easier
//
// ---------------
// There can be another way of writing error codes
// where we will write error codes with abbreviated feature
// for example if validation error occurs in user onboarding with agency
// we can write error code as UO-100 where UO- stands for user onboarding
// 1 stands for validation error, 0 stands which part of onboarding
// and last 0 stands for which parameter of part of onboarding
//
// There are advantages and disadvantages to all the approaches
//

export const CLAIM_STORAGE_ERROR = (e?: Error) => ({
  code: 'C-20001',
  message: `Failed to store claim ${e ? e.message : ''}`,
})

export const CONNECTION_HISTORY_STORAGE_ERROR = (e?: Error) => ({
  code: 'C-20002',
  message: `Failed to store connection history ${e ? e.message : ''}`,
})
