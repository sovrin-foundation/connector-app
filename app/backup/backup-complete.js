// @flow

import React, { PureComponent } from 'react'
import { Image, Dimensions } from 'react-native'
import { createStackNavigator } from 'react-navigation'
import {
  Container,
  CustomView,
  CustomText,
  Icon,
  CustomButton,
  CustomHeader,
} from '../components'
import { backupCompleteRoute } from '../common'
import {
  isBiggerThanVeryShortDevice,
  isBiggerThanShortDevice,
} from '../common/styles'
import { color } from '../common/styles/constant'
import type { BackupCompleteProps, BackupCompleteState } from './type-backup'
import {
  BACKUP_COMPLETE_CLOSE_TEST_ID,
  BACKUP_COMPLETE_SUBMIT_BUTTON,
  BACKUP_COMPLETE_SUBMIT_BUTTON_TITLE,
} from './backup-constants'
import styles from './styles'

const transparentBands = require('../images/transparentBands.png')
const closeImage = require('../images/iconClose.png')
// TODO: use real asset
const successCheck = require('../images/successCheck.png')

export class BackupComplete extends PureComponent<BackupCompleteProps, void> {
  backupComplete = () => {
    const { navigation: { navigate, state, goBack } } = this.props

    return navigate(state.params.initialRoute)
  }

  static navigationOptions = ({ navigation: { navigate, state } }) => ({
    header: (
      <CustomHeader
        backgroundColor={color.bg.fourteenth.color}
        largeHeader
        flatHeader
      >
        <CustomView style={[styles.headerSpacer]} />

        <CustomView style={[styles.headerSpacer]}>
          <Icon
            medium
            onPress={() => navigate(state.params.initialRoute)}
            testID={BACKUP_COMPLETE_CLOSE_TEST_ID}
            iconStyle={[styles.headerIcon]}
            src={closeImage}
          />
        </CustomView>
      </CustomHeader>
    ),
    gesturesEnabled: false,
  })

  render() {
    return (
      <Container style={[styles.backupComplete]} safeArea>
        <Image source={transparentBands} style={[styles.backgroundImage]} />
        <Container style={[styles.wrapper]}>
          <Container>
            <CustomView center>
              <CustomText transparentBg center style={[styles.title]}>
                Backup complete
              </CustomText>
            </CustomView>
            <CustomView center>
              <Image source={successCheck} />
            </CustomView>
            <CustomView horizontalSpace>
              <CustomView
                verticalSpace
                doubleVerticalSpace={isBiggerThanVeryShortDevice ? true : false}
              >
                <CustomText
                  transparentBg
                  center
                  style={styles.backupCompleteText}
                >
                  If you ever have to start with a new installation of
                  Connect.Me you will need to recover from this saved backup
                  file.
                </CustomText>
              </CustomView>
              <CustomView
                doubleVerticalSpace={isBiggerThanVeryShortDevice ? true : false}
              >
                <CustomText
                  transparentBg
                  center
                  style={styles.backupCompleteText}
                >
                  You will be asked to enter your Recovery Phrase during setup
                  of your new Connect.Me application.
                </CustomText>
              </CustomView>
            </CustomView>
          </Container>
          <CustomButton
            large={isBiggerThanShortDevice ? true : false}
            onPress={this.backupComplete}
            testID={BACKUP_COMPLETE_SUBMIT_BUTTON}
            style={[styles.submitButton]}
            customColor={{
              color: color.bg.fourteenth.color,
              fontWeight: '600',
              fontSize: 18,
            }}
            title={BACKUP_COMPLETE_SUBMIT_BUTTON_TITLE}
          />
        </Container>
      </Container>
    )
  }
}

export default createStackNavigator({
  [backupCompleteRoute]: {
    screen: BackupComplete,
  },
})
