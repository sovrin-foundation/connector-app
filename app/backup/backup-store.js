// @flow

import {
  put,
  takeLatest,
  call,
  all,
  select,
  fork,
  take,
  race,
} from 'redux-saga/effects'
import { delay } from 'redux-saga'
import { zip } from 'react-native-zip-archive'
import {
  GENERATE_RECOVERY_PHRASE_SUCCESS,
  GENERATE_BACKUP_FILE_SUCCESS,
  GENERATE_BACKUP_FILE_FAILURE,
  EXPORT_BACKUP_SUCCESS,
  EXPORT_BACKUP_FAILURE,
  BACKUP_COMPLETE,
  BACKUP_STORE_STATUS,
  BACKUP_WALLET_FAIL,
  PROMPT_WALLET_BACKUP_BANNER,
  GENERATE_RECOVERY_PHRASE_FAILURE,
  GENERATE_RECOVERY_PHRASE_LOADING,
  GENERATE_BACKUP_FILE_LOADING,
  PREPARE_BACKUP_LOADING,
  PREPARE_BACKUP_SUCCESS,
  PREPARE_BACKUP_FAILURE,
  EXPORT_BACKUP_LOADING,
  ERROR_EXPORT_BACKUP,
  ERROR_GENERATE_RECOVERY_PHRASE,
  ERROR_GENERATE_BACKUP_FILE,
  ERROR_HYDRATING_BACKUP,
  HYDRATE_BACKUP,
  HYDRATE_BACKUP_FAILURE,
  EXPORT_BACKUP_NO_SHARE,
  WALLET_FILE_NAME,
  PREPARE_BACK_IDLE,
} from './type-backup'
import type {
  PromptBackupBannerAction,
  BackupStore,
  BackupStoreStatus,
  BackupStoreAction,
  Passphrase,
  PrepareBackupLoadingAction,
  ExportBackupLoadingAction,
  GenerateRecoveryPhraseLoadingAction,
  GenerateBackupFileLoadingAction,
  PrepareBackupStatus,
} from './type-backup'
import RNFetchBlob from 'react-native-fetch-blob'
import { Platform } from 'react-native'
import Share from 'react-native-share'
import type { Saga } from 'redux-saga'
import {
  secureSet,
  secureGet,
  secureGetAll,
  secureDelete,
  safeSet,
  safeGet,
  walletSet,
  walletGet,
} from '../services/storage'
import type { AgencyPoolConfig } from '../store/type-config-store'
import type { CustomError } from '../common/type-common'
import {
  LAST_SUCCESSFUL_BACKUP,
  PASSPHRASE_SALT_STORAGE_KEY,
  PASSPHRASE_STORAGE_KEY,
} from '../common/secure-storage-constants'
import { WALLET_KEY } from '../bridge/react-native-cxs/vcx-transformers'
import { encryptWallet } from '../bridge/react-native-cxs/RNCxs'
import {
  getSalt,
  getConfig,
  getBackupPassphrase,
  getBackupWalletPath,
  getVcxInitializationState,
  getPrepareBackupStatus,
} from '../store/store-selector'
import { STORAGE_KEY_SHOW_BANNER } from '../components/banner/banner-constants'
import { getWords } from './secure-passphrase'
import { VCX_INIT_SUCCESS, __uniqueId } from '../store/type-config-store'
import { pinHash as generateKey, generateSalt } from '../lock/pin-hash'
import moment from 'moment'
import { captureError } from '../services/error/error-handler'

const initialState = {
  passphrase: { phrase: '', salt: 's', hash: 'h' },
  status: BACKUP_STORE_STATUS.IDLE,
  error: null,
  showBanner: false,
  lastSuccessfulBackup: '',
  backupWalletPath: '',
  prepareBackupStatus: PREPARE_BACK_IDLE,
}

export function* generateBackupSaga(
  action: GenerateBackupFileLoadingAction
): Generator<*, *, *> {
  // WALLET BACKUP ZIP FLOW
  const recoveryPassphrase: Passphrase = yield select(getBackupPassphrase)
  const { fs } = RNFetchBlob
  try {
    yield put(prepareBackup())

    const documentDirectory: string = fs.dirs.DocumentDir
    const backupTimeStamp = moment().format('YYYY-MM-DD-HH-mm-ss')
    const zipupDirectory: string = `${documentDirectory}/Backup-${backupTimeStamp}`

    // delete zip up directory if it already exists
    const destFileExists = yield call(fs.exists, zipupDirectory)
    if (destFileExists) {
      yield call(fs.unlink, zipupDirectory)
    }

    // create a new zip up directory
    yield call(fs.mkdir, zipupDirectory)
    const destinationZipPath: string = `${documentDirectory}/${WALLET_FILE_NAME}-${backupTimeStamp}.zip`
    const encryptedFileName: string = `${WALLET_FILE_NAME}.wallet`
    const encryptedFileLocation: string = `${zipupDirectory}/${encryptedFileName}`

    // create a file which contains salt needed to decrypt the wallet inside of zip up directory
    const saltValue = yield select(getSalt)
    const saltFileContents = JSON.stringify({ salt: saltValue })
    const saltFileName = `${zipupDirectory}/salt.json`

    const saltFile = yield call(
      fs.createFile,
      saltFileName,
      saltFileContents,
      'utf8'
    )

    // check status for prepare backup, with status of prepare backup
    // we would know whether we have moved all of data inside wallet or not
    // only when we are sure that we have moved data inside of wallet
    // we will go ahead and get a backup of wallet
    const prepareBackupStatus: PrepareBackupStatus = yield select(
      getPrepareBackupStatus
    )
    if (prepareBackupStatus !== PREPARE_BACKUP_SUCCESS) {
      // prepare backup can only be in 3 states, loading, success, failure
      // if it is not success, then we check for failure
      if (prepareBackupStatus === PREPARE_BACKUP_FAILURE) {
        throw new Error('Failed to write data back to wallet')
      }
      // if status is neither success nor failure, then we wait for it to success

      // we will wait for 2 minutes for data to go into wallet
      // and then we will timeout
      const { prepareBackupSuccess, timeout } = yield race({
        prepareBackupSuccess: take(PREPARE_BACKUP_SUCCESS),
        timeout: call(delay, 120000),
      })
      if (timeout) {
        throw new Error('Could not write data back to wallet in 2 minutes')
      }
    }

    // call VCX method to encrypt wallet. Given path to store it, file name, & key to encrypt it
    yield call(encryptWallet, {
      encryptedFileLocation,
      recoveryPassphrase,
    })

    // create zip file of zip up directory contents to be used as backup
    const backupZipFile = yield call(zip, zipupDirectory, destinationZipPath)
    yield put(generateBackupFileSuccess(backupZipFile))
  } catch (e) {
    captureError(e)
    yield put(
      generateBackupFileFail({
        ...ERROR_GENERATE_BACKUP_FILE,
        message: `${ERROR_GENERATE_BACKUP_FILE.message} ${e.message}`,
      })
    )
  }
}

const resolvedPromise = () => Promise.resolve()

export function* prepareBackupSaga(
  action: PrepareBackupLoadingAction
): Generator<*, *, *> {
  const skipItems = [__uniqueId, WALLET_KEY]

  try {
    // get all items saved in secure storage
    const secureStorage = yield call(secureGetAll)
    // for Android secureGetAll returns an object
    // while for ios it returns an array of array
    if (Platform.OS === 'android') {
      yield all(
        Object.keys(secureStorage).map(key => {
          if (skipItems.indexOf(key) === -1) {
            return call(walletSet, key, secureStorage[key])
          }

          return call(resolvedPromise)
        })
      )
    } else {
      if (secureStorage.length > 0) {
        yield all(
          secureStorage[0].map(item => {
            const { key, value } = item
            // check for things we don't want stored in the wallet
            if (skipItems.indexOf(key) === -1) {
              // store items into wallet
              return call(walletSet, key, value)
            }

            return call(resolvedPromise)
          })
        )
      }
    }

    yield put(prepareBackupSuccess())
  } catch (e) {
    yield put(
      prepareBackupFail({
        ...ERROR_GENERATE_BACKUP_FILE,
        message: `${ERROR_GENERATE_BACKUP_FILE.message} ${e.message}`,
      })
    )
  }
}

export function* exportBackupSaga(
  action: ExportBackupLoadingAction
): Generator<*, *, *> {
  try {
    const backupWalletPath: string = yield select(getBackupWalletPath)
    const title: string = `Export ${backupWalletPath.split('/').pop()}`
    Platform.OS === 'android'
      ? yield call(Share.open, {
          title,
          url: `file://${backupWalletPath}`,
          type: 'application/zip',
        })
      : yield call(Share.open, {
          title,
          url: backupWalletPath,
          type: 'application/zip',
        })
    const lastSuccessfulBackup = moment().format()
    yield put(exportBackupSuccess(lastSuccessfulBackup))
    yield put(promptBackupBanner(false))
    yield call(safeSet, LAST_SUCCESSFUL_BACKUP, lastSuccessfulBackup)
  } catch (e) {
    if (e.message === 'User did not share') {
      yield put(exportBackupNoShare())
    } else {
      captureError(new Error(`${ERROR_EXPORT_BACKUP.message} ${e.message}`))
      yield put(
        exportBackupFail({
          ...ERROR_EXPORT_BACKUP,
          message: `${ERROR_EXPORT_BACKUP.message} ${e.message}`,
        })
      )
    }
  }
}

export function* deletePersistedPassphrase(): Generator<*, *, *> {
  yield all([
    call(secureDelete, PASSPHRASE_STORAGE_KEY),
    call(secureDelete, PASSPHRASE_SALT_STORAGE_KEY),
  ])
}

export function* hydratePassphraseFromWallet(): Generator<*, *, *> {
  try {
    const [passphrase, passphraseSalt] = yield all([
      call(walletGet, PASSPHRASE_STORAGE_KEY),
      call(walletGet, PASSPHRASE_SALT_STORAGE_KEY),
    ])
    yield all([
      call(secureSet, PASSPHRASE_STORAGE_KEY, passphrase),
      call(secureSet, PASSPHRASE_SALT_STORAGE_KEY, passphraseSalt),
    ])
  } catch (e) {
    // not sure what to do if we don't get passphrase data
    // user would see new passphrase next time user is generating a backup
  }
}

export function* generateRecoveryPhraseSaga(
  action: GenerateRecoveryPhraseLoadingAction
): Generator<*, *, *> {
  const vcxInitializedState = yield select(getVcxInitializationState)
  if (vcxInitializedState !== VCX_INIT_SUCCESS) {
    yield take(VCX_INIT_SUCCESS)
  }
  // If it failed then we'll retry it a few times
  let retryCount = 0
  let lastInitException = new Error('')
  while (retryCount < 4) {
    try {
      let passphrase = yield call(secureGet, PASSPHRASE_STORAGE_KEY)
      let passphraseSalt = yield call(secureGet, PASSPHRASE_SALT_STORAGE_KEY)
      if (!passphrase) {
        const words: string[] = yield call(getWords, 8, 5)
        passphrase = words.join(' ')
        passphraseSalt = yield call(generateSalt)
      }
      const hashedPassphrase = yield call(
        generateKey,
        passphrase,
        passphraseSalt
      )
      yield call(secureSet, PASSPHRASE_STORAGE_KEY, passphrase)
      yield call(secureSet, PASSPHRASE_SALT_STORAGE_KEY, passphraseSalt)
      yield put(
        generateRecoveryPhraseSuccess({
          phrase: passphrase,
          salt: passphraseSalt,
          hash: hashedPassphrase,
        })
      )
      yield put(generateBackupFile())
      break
    } catch (e) {
      lastInitException = e
      retryCount++
    }
  }
  if (retryCount > 3) {
    captureError(new Error(`${ERROR_GENERATE_RECOVERY_PHRASE.message}`))
    yield put(
      generateRecoveryPhraseFail({
        ...ERROR_GENERATE_RECOVERY_PHRASE,
        message: `${ERROR_GENERATE_RECOVERY_PHRASE.message}`,
      })
    )
  }
}

export const prepareBackupSuccess = () => ({
  type: PREPARE_BACKUP_SUCCESS,
  status: BACKUP_STORE_STATUS.PREPARE_BACKUP_SUCCESS,
})

export const prepareBackupFail = (error: CustomError) => ({
  type: PREPARE_BACKUP_FAILURE,
  status: BACKUP_STORE_STATUS.PREPARE_BACKUP_FAILURE,
  error,
})

export const generateBackupFileSuccess = (backupWalletPath: string) => ({
  type: GENERATE_BACKUP_FILE_SUCCESS,
  status: BACKUP_STORE_STATUS.GENERATE_BACKUP_FILE_SUCCESS,
  backupWalletPath,
})

export const generateBackupFileFail = (error: CustomError) => ({
  type: GENERATE_BACKUP_FILE_FAILURE,
  status: BACKUP_STORE_STATUS.GENERATE_BACKUP_FILE_FAILURE,
  error,
})

export const hydrateBackupFail = (error: CustomError) => ({
  type: HYDRATE_BACKUP_FAILURE,
  error,
})

export const generateBackupFile = () => ({
  type: GENERATE_BACKUP_FILE_LOADING,
  status: BACKUP_STORE_STATUS.GENERATE_BACKUP_FILE_LOADING,
})

function* watchBackupStart(): any {
  yield takeLatest(GENERATE_BACKUP_FILE_LOADING, generateBackupSaga)
}

function* watchPrepareBackup(): any {
  yield takeLatest(PREPARE_BACKUP_LOADING, prepareBackupSaga)
}

function* watchExportBackup(): any {
  yield takeLatest(EXPORT_BACKUP_LOADING, exportBackupSaga)
}

function* watchGenerateRecoveryPhrase(): any {
  yield takeLatest(GENERATE_RECOVERY_PHRASE_LOADING, generateRecoveryPhraseSaga)
}

export function* watchBackup(): any {
  yield all([
    watchBackupStart(),
    watchPrepareBackup(),
    watchGenerateRecoveryPhrase(),
    watchExportBackup(),
    watchBackupBannerPrompt(),
  ])
}

export function* watchBackupBannerPrompt(): any {
  yield takeLatest(PROMPT_WALLET_BACKUP_BANNER, persistBackupBannerStatusSaga)
}

export function* hydrateBackupSaga(): Generator<*, *, *> {
  // We run two hydrations that are needed by this store in single saga
  // we could have merged these two into single hydration and followed
  // one hydration per reducer, for now we are just moving ahead with two
  // Also, these two hydrations runs in parallel, and even if one hydration fails
  // other hydration will continue
  // reason for not using "all" effect of redux-saga is that in all if one saga fails
  // then whole "all" effect is cancelled and other saga are not run to completion
  yield fork(hydrateLastSuccessfulBackupSaga)
  yield fork(hydrateBackupBannerStatusSaga)
}

export function* hydrateLastSuccessfulBackupSaga(): Generator<*, *, *> {
  try {
    let lastSuccessfulBackup = yield call(safeGet, LAST_SUCCESSFUL_BACKUP)
    if (lastSuccessfulBackup != null) {
      yield put(hydrateBackup(lastSuccessfulBackup))
    }
  } catch (e) {
    captureError(e)
    yield put(
      hydrateBackupFail({
        ...ERROR_HYDRATING_BACKUP,
        message: `${ERROR_HYDRATING_BACKUP.message} ${e.message}`,
      })
    )
  }
}

export function* hydrateBackupBannerStatusSaga(): Generator<*, *, *> {
  try {
    const backupBannerStatus = yield call(safeGet, STORAGE_KEY_SHOW_BANNER)
    if (backupBannerStatus) {
      yield put(promptBackupBanner(JSON.parse(backupBannerStatus)))
    }
  } catch (e) {
    captureError(e)
    yield put(
      hydrateBackupFail({
        ...ERROR_HYDRATING_BACKUP,
        message: `${ERROR_HYDRATING_BACKUP.message} ${e.message}`,
      })
    )
  }
}

export function* persistBackupBannerStatusSaga(
  action: PromptBackupBannerAction
): Generator<*, *, *> {
  const { showBanner } = action
  try {
    yield call(safeSet, STORAGE_KEY_SHOW_BANNER, JSON.stringify(showBanner))
  } catch (e) {
    yield put(promptBackupBanner(showBanner))
  }
}

export const hydrateBackup = (lastSuccessfulBackup: string) => ({
  type: HYDRATE_BACKUP,
  lastSuccessfulBackup,
})

export const promptBackupBanner = (
  showBanner: boolean
): PromptBackupBannerAction => ({
  type: PROMPT_WALLET_BACKUP_BANNER,
  showBanner,
})

export const generateRecoveryPhrase = () => ({
  type: GENERATE_RECOVERY_PHRASE_LOADING,
  status: BACKUP_STORE_STATUS.GENERATE_PHRASE_LOADING,
})

export const generateRecoveryPhraseSuccess = (passphrase: Passphrase) => ({
  type: GENERATE_RECOVERY_PHRASE_SUCCESS,
  status: BACKUP_STORE_STATUS.GENERATE_PHRASE_SUCCESS,
  passphrase: passphrase,
})

export const generateRecoveryPhraseFail = (error: CustomError) => ({
  type: GENERATE_RECOVERY_PHRASE_FAILURE,
  status: BACKUP_STORE_STATUS.GENERATE_BACKUP_FILE_FAILURE,
  error,
})

export const prepareBackup = () => ({
  type: PREPARE_BACKUP_LOADING,
  status: BACKUP_STORE_STATUS.PREPARE_BACKUP_LOADING,
})

export const exportBackup = () => ({
  type: EXPORT_BACKUP_LOADING,
  status: BACKUP_STORE_STATUS.EXPORT_BACKUP_LOADING,
})

export const exportBackupFail = (error: CustomError) => ({
  type: EXPORT_BACKUP_FAILURE,
  error,
  status: BACKUP_STORE_STATUS.EXPORT_BACKUP_FAILURE,
})

export const exportBackupSuccess = (lastSuccessfulBackup: string) => ({
  type: EXPORT_BACKUP_SUCCESS,
  status: BACKUP_COMPLETE,
  lastSuccessfulBackup,
})

export const exportBackupNoShare = () => ({
  type: EXPORT_BACKUP_NO_SHARE,
  status: BACKUP_STORE_STATUS.EXPORT_BACKUP_NO_SHARE,
})

export default function backupReducer(
  state: BackupStore = initialState,
  action: BackupStoreAction
) {
  switch (action.type) {
    case GENERATE_RECOVERY_PHRASE_LOADING: {
      return {
        ...state,
        status: action.status,
        error: null,
      }
    }
    case GENERATE_RECOVERY_PHRASE_SUCCESS: {
      return {
        ...state,
        status: action.status,
        error: action.error,
        passphrase: {
          phrase: action.passphrase.phrase,
          salt: action.passphrase.salt,
          hash: action.passphrase.hash,
        },
      }
    }
    case GENERATE_RECOVERY_PHRASE_FAILURE: {
      return {
        ...state,
        status: action.status,
        error: action.error,
      }
    }
    case GENERATE_BACKUP_FILE_LOADING: {
      return {
        ...state,
        status: action.status,
        error: null,
      }
    }
    case GENERATE_BACKUP_FILE_SUCCESS: {
      return {
        ...state,
        status: action.status,
        error: null,
        backupWalletPath: action.backupWalletPath,
      }
    }
    case GENERATE_BACKUP_FILE_FAILURE: {
      return {
        ...state,
        status: action.status,
        error: action.error,
      }
    }
    case EXPORT_BACKUP_LOADING: {
      return {
        ...state,
        status: action.status,
        error: null,
      }
    }
    case EXPORT_BACKUP_NO_SHARE: {
      return {
        ...state,
        status: action.status,
        error: null,
      }
    }
    case EXPORT_BACKUP_SUCCESS: {
      return {
        ...state,
        error: null,
        status: action.status,
        lastSuccessfulBackup: action.lastSuccessfulBackup,
      }
    }
    case EXPORT_BACKUP_FAILURE: {
      return {
        ...state,
        status: action.status,
        error: action.error,
      }
    }
    case PROMPT_WALLET_BACKUP_BANNER: {
      return {
        ...state,
        showBanner: action.showBanner,
      }
    }
    case BACKUP_WALLET_FAIL: {
      return {
        ...state,
        error: action.error,
      }
    }
    case HYDRATE_BACKUP:
      return {
        ...initialState,
        lastSuccessfulBackup: action.lastSuccessfulBackup,
        showBanner: state.showBanner,
      }
    case HYDRATE_BACKUP_FAILURE:
      return {
        ...state,
        error: action.error,
      }
    case PREPARE_BACKUP_LOADING:
    case PREPARE_BACKUP_SUCCESS:
    case PREPARE_BACKUP_FAILURE:
      return {
        ...state,
        prepareBackupStatus: action.status,
      }
    default:
      return state
  }
}
