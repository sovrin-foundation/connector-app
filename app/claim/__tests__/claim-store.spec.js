// @flow
import renderer from 'react-test-renderer'
import { put, call, select } from 'redux-saga/effects'
import { initialTestAction } from '../../common/type-common'
import claimReducer, {
  claimReceived,
  claimStorageFail,
  claimStorageSuccess,
  mapClaimToSender,
} from '../claim-store'
import { CLAIM_STORAGE_ERROR } from '../../services/error/error-code'
import {
  getConnectionLogoUrl,
  getPoolConfig,
  getClaimMap,
} from '../../store/store-selector'
import {
  claim,
  claimWithUuid,
  senderDid1,
  claimDefinitionSchemaSequenceNumber,
  getClaimFormat,
  senderLogoUrl1,
  myPairWiseConnectionDetails,
  poolConfig,
  claimMap,
} from '../../../__mocks__/static-data'
import type { ClaimWithUuid } from '../../claim/type-claim'
import { promptBackupBanner } from '../../backup/backup-store'

describe('Claim Store', () => {
  let initialState = { claimMap: {} }
  let afterClaimReceived

  beforeEach(() => {
    initialState = claimReducer(undefined, initialTestAction())
    afterClaimReceived = claimReducer(initialState, claimReceived(claim))
  })

  it('should match snapshot for claim received action', () => {
    expect(afterClaimReceived).toMatchSnapshot()
  })

  it('should match snapshot when claim storage fails', () => {
    const nextState = claimReducer(
      afterClaimReceived,
      claimStorageFail(claim.messageId, CLAIM_STORAGE_ERROR())
    )
    expect(nextState).toMatchSnapshot()
  })

  it('should match snapshot when claim storage is success', () => {
    const nextState = claimReducer(
      afterClaimReceived,
      claimStorageSuccess(claim.messageId)
    )
    expect(nextState).toMatchSnapshot()
  })

  it('should reset claim store, if RESET action is raised', () => {
    expect(
      claimReducer(afterClaimReceived, { type: 'RESET' })
    ).toMatchSnapshot()
  })
})
