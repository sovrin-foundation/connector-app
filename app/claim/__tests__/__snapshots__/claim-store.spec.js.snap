// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Claim Store should match snapshot for claim received action 1`] = `
Object {
  "1": Object {
    "claim": Object {
      "claim": Object {
        "date_of_birth": Array [
          "20-2-1800",
          "anon cred date",
        ],
        "name": Array [
          "test",
          "anon cred test",
        ],
      },
      "forDID": "forDID",
      "from_did": "from_did",
      "issuer_did": "issuerDid",
      "messageId": "1",
      "remoteDid": "remoteDid",
      "schema_seq_no": 36,
      "signature": Object {
        "primary_claim": Object {
          "a": "a",
          "e": "e",
          "m2": "m2",
          "v": "v",
        },
      },
      "uid": "usd123",
    },
  },
  "claimMap": Object {},
}
`;

exports[`Claim Store should match snapshot when claim storage fails 1`] = `
Object {
  "1": Object {
    "claim": Object {
      "claim": Object {
        "date_of_birth": Array [
          "20-2-1800",
          "anon cred date",
        ],
        "name": Array [
          "test",
          "anon cred test",
        ],
      },
      "forDID": "forDID",
      "from_did": "from_did",
      "issuer_did": "issuerDid",
      "messageId": "1",
      "remoteDid": "remoteDid",
      "schema_seq_no": 36,
      "signature": Object {
        "primary_claim": Object {
          "a": "a",
          "e": "e",
          "m2": "m2",
          "v": "v",
        },
      },
      "uid": "usd123",
    },
    "error": Object {
      "code": "C-20001",
      "message": "Failed to store claim ",
    },
  },
  "claimMap": Object {},
}
`;

exports[`Claim Store should match snapshot when claim storage is success 1`] = `
Object {
  "claimMap": Object {},
}
`;

exports[`Claim Store should reset claim store, if RESET action is raised 1`] = `
Object {
  "claimMap": Object {},
}
`;
